File: src/HotGoss/Challenge3a.hs
module HotGoss.Challenge3a (main) where

import Data.Aeson (FromJSON, ToJSON)
import Data.Data (Data)
import HotGoss.Protocol
import HotGoss.Union
import Prelude hiding (Read, on)

data Broadcast = Broadcast
  { msgId :: MessageId
  , message :: Word
  }
  deriving stock (Generic, Data, Show)
  deriving (ToJSON, FromJSON) via MessageBodyJson Broadcast

data BroadcastOk = BroadcastOk
  { msgId :: MessageId
  , inReplyTo :: MessageId
  }
  deriving stock (Generic, Data, Show)
  deriving (ToJSON, FromJSON) via MessageBodyJson BroadcastOk

data Read = Read
  { msgId :: MessageId
  }
  deriving stock (Generic, Data, Show)
  deriving (ToJSON, FromJSON) via MessageBodyJson Read

data ReadOk = ReadOk
  { msgId :: MessageId
  , inReplyTo :: MessageId
  , messages :: [Word]
  }
  deriving stock (Generic, Data, Show)
  deriving (ToJSON, FromJSON) via MessageBodyJson ReadOk

data Topology = Topology
  { msgId :: MessageId
  , topology :: HashMap NodeId [NodeId]
  }
  deriving stock (Generic, Data, Show)
  deriving (ToJSON, FromJSON) via MessageBodyJson Topology

data TopologyOk = TopologyOk
  { msgId :: MessageId
  , inReplyTo :: MessageId
  }
  deriving stock (Generic, Data, Show)
  deriving (ToJSON, FromJSON) via MessageBodyJson TopologyOk

main :: IO ()
main = do
  messagesRef <- newIORef []

  (getMessageId, _, _) <- handleInit

  handle @Topology \body -> do
    -- TODO: body.topology
    msgId <- getMessageId
    pure TopologyOk
      { msgId
      , inReplyTo = body.msgId
      }

  let handleBroadcast :: Broadcast -> IO BroadcastOk
      handleBroadcast body = do
        msgId <- getMessageId
        atomicModifyIORef' messagesRef \ms -> (body.message : ms, ())
        pure BroadcastOk
          { msgId
          , inReplyTo = body.msgId
          }

  let handleRead :: Read -> IO ReadOk
      handleRead body = do
        msgId <- getMessageId
        messages <- readIORef messagesRef
        pure ReadOk
          { msgId
          , inReplyTo = body.msgId
          , messages
          }

  forever $ handle @_ @(Union '[BroadcastOk, ReadOk]) $
    case_
      `on` (\msg -> handleRead msg <&> inject)
      `on` (\msg -> handleBroadcast msg <&> inject)

File: src/challenge_3a.rs
use crate::protocol::{handle, handle_init, Either, Message, MessageId, NodeId};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Deserialize)]
#[serde(rename = "broadcast", tag = "type")]
struct Broadcast {
    msg_id: MessageId,
    message: usize,
}

#[derive(Serialize)]
#[serde(rename = "broadcast_ok", tag = "type")]
struct BroadcastOk {
    msg_id: MessageId,
    in_reply_to: MessageId,
}

#[derive(Deserialize)]
#[serde(rename = "read", tag = "type")]
struct Read {
    msg_id: MessageId,
}

#[derive(Serialize)]
#[serde(rename = "read_ok", tag = "type")]
struct ReadOk {
    msg_id: MessageId,
    in_reply_to: MessageId,
    messages: Vec<usize>,
}

#[derive(Deserialize)]
#[serde(rename = "topology", tag = "type")]
struct Topology {
    msg_id: MessageId,
    #[allow(dead_code)]
    topology: HashMap<NodeId, Vec<NodeId>>,
}

#[derive(Serialize)]
#[serde(rename = "topology_ok", tag = "type")]
struct TopologyOk {
    msg_id: MessageId,
    in_reply_to: MessageId,
}

pub fn main() -> anyhow::Result<()> {
    let mut messages = Vec::new();

    let (mut msg_id, _, _) = handle_init()?;

    handle(|request: Message<Topology>| {
        Ok(TopologyOk {
            msg_id: msg_id.next(),
            in_reply_to: request.body.msg_id,
        })
    })?;

    loop {
        handle(|request: Message<Either<Broadcast, Read>>| {
            Ok(match request.body {
                Either::Left(body) => {
                    messages.push(body.message);
                    Either::Left(BroadcastOk {
                        msg_id: msg_id.next(),
                        in_reply_to: body.msg_id,
                    })
                }
                Either::Right(body) => Either::Right(ReadOk {
                    msg_id: msg_id.next(),
                    in_reply_to: body.msg_id,
                    messages: messages.clone(),
                }),
            })
        })?;
    }
}
